subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java

```diff
@@ -192,22 +191,30 @@ private V produce(Function<? super K, ? extends V> factory, K k) {
         }
     }
 
+    /**
+     * Specify the retention of keys in {@link DefaultCrossBuildInMemoryCache}.
+     */
     private enum KeyRetentionPolicy {
         WEAK,
-        STRONG
+        STRONG,
     }
 
     private static class DefaultCrossBuildInMemoryCache<K, V> extends AbstractCrossBuildInMemoryCache<K, V> {
 
-        // This is used only to retain strong references to the values
-        private final Set<V> valuesForPreviousSession = new HashSet<>();
-        private final Map<K, SoftReference<V>> allValues;
+        /**
+         * This is used only to retain weak references to the values.
+         * Use weak references, in case Java needs to GC the values.
+         */
+        private final Set<V> valuesForPreviousSession = newSetFromMap(new WeakHashMap<>());
```

Comment:
I changed `valuesForPreviousSession` from hard to soft references, because if Java needs to GC the values then I don't think Gradle should prevent that!

The Javadoc in the parent interfaces is aligned with the new behaviour in this PR:

https://github.com/gradle/gradle/blob/28326f4cf6bf430e725480711b2993058859371b/platforms/core-execution/persistent-cache/src/main/java/org/gradle/cache/internal/ClassCacheFactory.java#L34-L35